London | 25-SDC-Nov | Emiliano Uruena | Sprint 5 |  Laptops Allocation

Emilianouz opened the pull request:
Self checklist

- [x] I have titled my PR with Region | Cohort | FirstName LastName | Sprint | Assignment Title
- [x] My changes meet the requirements of the task
- [x] I have tested my changes
- [x] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/reviewing/style-guide/)

## Changelist

Implementing Laptop allocation exercise.

Emilianouz committed:
laptop allocation exercise

OracPrime commented on number-systems/README.md:
```diff
@@ -5,61 +5,62 @@ Do not convert any binary numbers to decimal when solving a question unless the
 The goal of these exercises is for you to gain an intuition for binary numbers. Using tools to solve the problems defeats the point.
 
 Convert the decimal number 14 to binary.
-Answer:
```
What's this file doing in the laptops-allocation task?

OracPrime commented on laptop-allocation.py:
```diff
@@ -0,0 +1,62 @@
+    for person in people:
+        for os in person.preferred_operating_system:
+            for laptop in laptops:
+                if (laptop.operating_system == os and not laptop.assigned):
```
Have you run this code and checked the output?  Can you paste a copy as a comment so I can see it has worked?  Thanks

Emilianouz commented:
README file no necessary. It was removed.

SlideGauge commented on laptop-allocation.py:
```diff
@@ -0,0 +1,62 @@
+            for laptop in laptops:
+                if (laptop.operating_system == os and not laptop.assigned):
+                    allocation[person.name] = laptop.id
+                    assigned = True
```
Does this assigned variable prevent us from selecting the same laptop several times?

SlideGauge commented on laptop-allocation.py:
```diff
@@ -0,0 +1,62 @@
+                    allocation[person.name] = laptop.id
+                    assigned = True
+                    break
+            if assigned:
```
what happens if we did not found the laptop after the first loop
```for laptop in laptops```?

SlideGauge replied:
if assigned was set to True, what happens after iterating the first os?

SlideGauge commented on laptop-allocation.py:
```diff
@@ -0,0 +1,62 @@
+    operating_system: OperatingSystem
+    assigned: bool = False
+
+def allocate_laptops(people: List[Person], laptops: List[Laptop]) -> Dict[str, int]:
```
The algorithm belongs to the type of algorithms called "Greedy", i.e. on each step the algorithm takes the most suitable choice seen **at this moment**, not **globally**.
Consider:
Imran wants UBUNTU, ARCH
Eliza wants UBUNTU

The algorithm will give UBUNTU to Imran and nothing to Eliza, leading to sadness 100.

But we can invent different algorithm: if we distribute UBUNTU to Eliza and ARCH to Imran, we will have sadness 1. This algorithm will be slower, but our task is not to create the quickest algorithm, but minimizing the sadness.

Hint: this task requires analyzing of all possible permutations of laptops distribution.

SlideGauge commented on laptop-allocation.py:
```diff
@@ -0,0 +1,62 @@
+    name: str
+    age: int
+    preferred_operating_system: List[OperatingSystem]
+    assigned_laptop: Optional[int] = None
```
Since the dataclass is frozen, we won't be able to change the value of this variable

SlideGauge commented:
Made review, interesting approach  but there are some errors in the implementation itself worthy fixing + the algorithm won't find the most minimal possible configuration

Emilianouz committed:
Add sadness calculator and backtracking for laptop allocation

Implement calcutate_sandess and backtracking for laptop allocation based on user preferences and minimize sadness.

OracPrime commented:
I'll leave @SlideGauge to carry on with this since he's most recent reviewer

SlideGauge commented on laptop-allocation.py:
```diff
@@ -0,0 +1,96 @@
+        return person.preferred_operating_system.index(laptop.operating_system)
+    return 100
+
+def allocate_laptops(people: List[Person], laptops: List[Laptop]) -> Dict[str, int]:
```
Really nice, yes, backtrack will solve the task.

SlideGauge commented on laptop-allocation.py:
```diff
@@ -0,0 +1,96 @@
+        return person.preferred_operating_system.index(laptop.operating_system)
+    return 100
+
+def allocate_laptops(people: List[Person], laptops: List[Laptop]) -> Dict[str, int]:
```
One thing to take into account: 
the task demands us for the function to have this signature:
```def allocate_laptops(people: List[Person], laptops: List[Laptop]) -> Dict[Person, Laptop]:```
Note, we return Dict of  Person->Laptop, not str->int. It is valuable to take into account, as if we just return str->int, we won't be able to distinguish to people with the same name.

Emilianouz replied:
Right. We should return Dict[Person, Laptop] so people with the same name are handled correctly.

SlideGauge reviewed (commented):
Overall really good solution, please change the returning type of the function and I will mark the review as complete.

Emilianouz committed:
Change allocation function to change the returning type of the function: Person and Laptop objects

Change allocation function to change the returning type of the function: Person and Laptop objects

Emilianouz commented:
@SlideGauge , I changed it to return type of the function: Person and Laptop objects. Thank you

SlideGauge commented:
Very nice, thank you, accepted
